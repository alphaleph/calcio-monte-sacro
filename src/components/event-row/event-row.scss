@import '../../pages/globals.scss';

.event-row-container {
    min-width: 280px;
    min-height:250px;
    background-color: $white-main;
    color: $black-main;
    font-weight: bold;
    border-bottom: 1px $roma-gold solid;
}

.event-row-header {
    padding: 2rem;
}

.event-row-info {
    padding-bottom: 5rem;
}

.team-info-abrv {
    padding: 1rem;
}

.match-info-score {
    padding: 0rem;
}

.team-info {
    font-size: 1.2rem;
}

.match-info {
    font-size: 1.5rem;
    padding: 0rem;
}

.practice-info {
    font-size: 2.3rem;
}

.team-info-icon {
    font-size: 1rem;
}

@media screen and (min-width: 400px) {
    .team-info {
        font-size: 1.5rem;
    }
    .match-info {
        font-size: 2rem;
    }
    .team-info-icon {
        font-size: 1.75rem;
    }
}

@media screen and (min-width: 500px) {
    .team-info {
        font-size: 1.7rem;
    }
    .match-info {
        font-size: 2.5rem;
    }
    .team-info-icon {
        font-size: 3rem;
    }
}

@media screen and (min-width: 860px) {
    .team-info {
        font-size: 2rem;
    }
    .match-info {
        font-size: 3rem;
    }
    .match-info-score {
        padding: 1rem;
    }
    .team-info-icon {
        font-size: 3rem;
    }
}

@media screen and (min-width: 1023px) {
    .team-info {
        font-size: 2.5rem;
    }
    .match-info {
        font-size: 3.5rem;
    }
    .match-info-score {
        padding: 1rem;
    }
    .practice-info {
        font-size: 3.5rem;
    }
    .team-info-icon {
        font-size: 3rem;
    }
}