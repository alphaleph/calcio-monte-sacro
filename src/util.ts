import { MarkdownDataNode } from './types/index';

/**@func findMostRecentlyPastIndex
 * Finds index of most recently past Markdown post. Implemented via a modified binary search.
 * @param {MarkdownDataNode[]} edges Array of Markdown nodes. Assumes dates are sorted in ASC order!
 * @return {number} The index of the most recently past Markdown post. Returns -1 if current date is the earliest date.
 * @throws Error if edges size is 0.
 */
export const findMostRecentlyPastIndex = (
    edges: MarkdownDataNode[]
): number => {
    if (edges.length === 0) {
        throw Error('Error: There are no Markdown nodes for searching!');
    }

    const currDate = Date.now();

    // Case: Returns latest post if today is latest date
    // Case: Returns -1 if today is earliest date
    if (Date.parse(edges[edges.length - 1].node.frontmatter.date) < currDate) {
        return edges.length - 1;
    } else if (Date.parse(edges[0].node.frontmatter.date) > currDate) {
        return -1;
    }

    // Case: Return index of most recently past post
    const dates = edges.map((edge: MarkdownDataNode) =>
        Date.parse(edge.node.frontmatter.date)
    );

    return lowestNeighborIndexBinarySearch(dates, currDate);
};

/**@func lowestNeighborIndexBinarySearch
 * Finds via binary search: the index of the query value, of the nearest lower array value, or of the lowest array value if query value is the lowest.
 * @param {number[]} arr Array of numbers. Assumes dates are sorted in ASC order!
 * @param {number} val Value whose index to search for.
 * @return {number} The index of the query value. Otherwise, the index of the lowest neighbor.
 * @throws Error if array size is 0.
 */
export const lowestNeighborIndexBinarySearch = (
    arr: number[],
    val: number
): number => {
    if (arr.length === 0) {
        throw Error('Error: Array has size less than 1!');
    }

    let currIndex = Math.floor(arr.length / 2);
    let finalStep = Math.floor(Math.log2(arr.length)) - 1;
    let indexModifier = 0;

    for (let i = 0; i < Math.floor(Math.log2(arr.length)); i++) {
        if (arr[currIndex] === val) {
            return currIndex;
        }

        indexModifier = Math.max(
            Math.floor(Math.floor(arr.length / (2 * (i + 1))) / 2),
            1
        );

        if (arr[currIndex] < val) {
            //Look into upper half
            if (
                i < finalStep ||
                (arr.length % 2 === 0 && val >= arr[currIndex + indexModifier])
            ) {
                currIndex += indexModifier;
            }

            // Corner Case for two-off odd-sized arrays
            if (
                i === finalStep &&
                (arr.length % 2 === 1 && val >= arr[currIndex + indexModifier])
            ) {
                currIndex += indexModifier;
                // Inner Corner Case for one-off odd-sized arrays
                if (
                    currIndex !== arr.length &&
                    val >= arr[currIndex + indexModifier]
                ) {
                    currIndex += indexModifier;
                }
            }
        } else if (arr[currIndex] > val) {
            //Look into lower half
            currIndex -= indexModifier;
            // Corner Case due to one-off
            if (i === finalStep && (currIndex != 0 && val < arr[currIndex])) {
                currIndex -= indexModifier;
            }
        }
    }
    return currIndex;
};
