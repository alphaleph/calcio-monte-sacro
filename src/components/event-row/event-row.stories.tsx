import React from 'react';
import { storiesOf } from '@storybook/react';
import { EventRow } from './event-row';
import '../../pages/index.scss';
import {
    mockEventPracticeEN,
    mockEventPracticeIT,
    mockEventFinishedMatchEN,
    mockEventFinishedMatchIT,
    mockEventFutureMatchEN,
    mockEventFutureMatchIT,
} from '../__tests__/test-util';

storiesOf('EventRow', module)
    .add('practice-en', () => <EventRow event={mockEventPracticeEN} />)
    .add('match-finished-en', () => (
        <EventRow event={mockEventFinishedMatchEN} />
    ))
    .add('match-future-en', () => <EventRow event={mockEventFutureMatchEN} />)
    .add('practice-it', () => <EventRow event={mockEventPracticeIT} />)
    .add('match-finished-it', () => (
        <EventRow event={mockEventFinishedMatchIT} />
    ))
    .add('match-future-it', () => <EventRow event={mockEventFutureMatchIT} />)
    .add('empty', () => <EventRow event={null} />);
