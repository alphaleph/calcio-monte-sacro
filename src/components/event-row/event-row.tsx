import React from 'react';
import { Loading } from '../loading/loading';
import { utcStringToDateString } from '../../util';
import { MATCH_EN, MATCH_IT, HOME_VENUE, TEAM_ABRV } from '../../constants';
import { MarkdownEventDataNode } from '../../types';
import './event-row.scss';

interface EventRowProps {
    event: MarkdownEventDataNode | null;
}

export const EventRow = ({ event }: EventRowProps) => {
    let el = <Loading />;
    if (event) {
        const frontMatter = (event as MarkdownEventDataNode).node.frontmatter;
        const date = new Date(frontMatter.date);
        const venue = frontMatter.venue;
        const type = frontMatter.type;
        // Event: Match
        if (type === MATCH_EN || type === MATCH_IT) {
            const opposingTeam = frontMatter.opposing_team;
            let matchInfo = null;
            // Event: Finished Match
            if (date.getTime() < Date.now()) {
                const scoreMS = frontMatter.score_ms;
                const scoreOther = frontMatter.score_other;
                matchInfo = (
                    <p>
                        <span className="match-info-score">
                            {venue === HOME_VENUE ? scoreMS : scoreOther}
                        </span>
                        <span className="has-text-team-granite">-</span>
                        <span className="match-info-score">
                            {venue === HOME_VENUE ? scoreOther : scoreMS}
                        </span>
                    </p>
                );
            } else {
                matchInfo = <p>{`${date.getHours()}:${date.getMinutes()}`}</p>;
            }
            el = (
                <article className="event-row-container container">
                    <header className="event-row-header">
                        <p>{`${utcStringToDateString(frontMatter.date)}`}</p>
                        <p className="has-text-weight-light">{venue}</p>
                    </header>
                    <section className="event-row-info columns is-mobile is-vcentered">
                        <div className="column is-two-fifths has-text-right team-info">
                            <span className="team-info-abrv">
                                {venue === HOME_VENUE
                                    ? TEAM_ABRV
                                    : opposingTeam}
                            </span>
                            <span
                                className="team-info-icon"
                                role="img"
                                aria-label="Soccer ball"
                            >
                                ⚽
                            </span>
                        </div>
                        <div className="column is-one-fifth match-info has-text-roma-red has-text-centered">
                            {matchInfo}
                        </div>
                        <div className="column is-two-fifths team-info">
                            <span
                                className="team-info-icon"
                                role="img"
                                aria-label="Soccer ball"
                            >
                                ⚽
                            </span>
                            <span className="team-info-abrv">
                                {venue === HOME_VENUE
                                    ? opposingTeam
                                    : TEAM_ABRV}
                            </span>
                        </div>
                    </section>
                </article>
            );
        } else {
            // Event: Practice
            el = (
                <article className="event-row-container container">
                    <header className="event-row-header">
                        <p>{`${utcStringToDateString(frontMatter.date)}`}</p>
                        <p className="has-text-weight-light">{venue}</p>
                    </header>
                    <section className="event-row-info has-text-centered columns is-mobile is-vcentered">
                        <div className="column practice-info">
                            <p>PRACTICE</p>
                            <p className="has-text-roma-red">{`${date.getHours()}:${date.getMinutes()}`}</p>
                        </div>
                    </section>
                </article>
            );
        }
    } else {
        // Event: null
        el = (
            <article
                className="event-row-container container"
                data-testid="event-row-empty"
            />
        );
    }

    return el;
};
