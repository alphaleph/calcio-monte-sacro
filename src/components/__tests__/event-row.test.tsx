import React from 'react';
import { EventRow } from '../event-row/event-row';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import {
    mockEventPracticeEN,
    mockEventFinishedMatchEN,
    mockEventFutureMatchEN,
} from '../__tests__/test-util';
import { utcStringToDateString } from '../../util';

const VENUE = 'Parco Sannazzaro';
const OPP_TEAM = 'CMS2';
const SCORE_MS = 1;
const SCORE_OTHER = 2;
const DATE_FINISHED = utcStringToDateString('2019-11-20T17:40:15.578Z');
const rawDateFuture = new Date('2200-11-20T17:40:15.578Z');
const DATE_FUTURE = utcStringToDateString('2200-11-20T17:40:15.578Z');
const TIME = `${rawDateFuture.getHours()}:${rawDateFuture.getMinutes()} PM`;
const EMPTY_TEST_ID = 'event-row-empty';

//TODO: Error checks for incorrectly formatted input
//TODO: Test cases for non-home venues
//TODO: Replace team emblem placeholders
describe('Event Card', () => {
    it('renders without crashing--practice', () => {
        const { getByText } = render(<EventRow event={mockEventPracticeEN} />);
        expect(getByText(VENUE)).toBeInTheDocument();
    });

    it('renders without crashing--match-finished', () => {
        const { getByText } = render(
            <EventRow event={mockEventFinishedMatchEN} />
        );
        expect(getByText(VENUE)).toBeInTheDocument();
    });

    it('renders without crashing--match-future', () => {
        const { getByText } = render(
            <EventRow event={mockEventFutureMatchEN} />
        );
        expect(getByText(VENUE)).toBeInTheDocument();
    });

    it('renders without crashing--null', () => {
        const { getByTestId } = render(<EventRow event={null} />);
        expect(getByTestId(EMPTY_TEST_ID)).toBeInTheDocument();
    });

    it('renders correctly--practice', () => {
        const { getByText } = render(<EventRow event={mockEventPracticeEN} />);
        expect(getByText(VENUE)).toBeVisible();
        expect(getByText(DATE_FINISHED)).toBeVisible();
        expect(getByText(TIME)).toBeVisible();
        expect(getByText('PRACTICE')).toBeVisible();
    });

    it('renders correctly--match-finished', () => {
        const { getByText } = render(
            <EventRow event={mockEventFinishedMatchEN} />
        );
        expect(getByText(VENUE)).toBeVisible();
        expect(getByText(DATE_FINISHED)).toBeVisible();
        expect(getByText(OPP_TEAM)).toBeVisible();
        expect(getByText(`${SCORE_MS}`)).toBeVisible();
        expect(getByText(`${SCORE_OTHER}`)).toBeVisible();
    });

    it('renders correctly--match-future', () => {
        const { getByText } = render(
            <EventRow event={mockEventFutureMatchEN} />
        );
        expect(getByText(VENUE)).toBeVisible();
        expect(getByText(DATE_FUTURE)).toBeVisible();
        expect(getByText(TIME)).toBeVisible();
        expect(getByText(OPP_TEAM)).toBeVisible();
    });

    it('renders correctly--null', () => {
        const { getByTestId } = render(<EventRow event={null} />);
        expect(getByTestId(EMPTY_TEST_ID)).toBeVisible();
    });

    it('has not changed--practice', () => {
        const { asFragment } = render(<EventRow event={mockEventPracticeEN} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('has not changed--match-finished', () => {
        const { asFragment } = render(
            <EventRow event={mockEventFinishedMatchEN} />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('has not changed--match-future', () => {
        const { asFragment } = render(
            <EventRow event={mockEventFutureMatchEN} />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('has not changed--null', () => {
        const { asFragment } = render(<EventRow event={null} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
