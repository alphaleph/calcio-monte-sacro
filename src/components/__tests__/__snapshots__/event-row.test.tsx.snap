// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event Card has not changed--match-finished 1`] = `
<DocumentFragment>
  <article
    class="event-row-container container"
  >
    <header
      class="event-row-header"
    >
      <p>
        20 NOV 2019
      </p>
      <p
        class="has-text-weight-light"
      >
        Parco Sannazzaro
      </p>
    </header>
    <section
      class="event-row-info columns is-mobile is-vcentered"
    >
      <div
        class="column is-two-fifths has-text-right team-info"
      >
        <span
          class="team-info-abrv"
        >
          CMS
        </span>
        <span
          aria-label="Soccer ball"
          class="team-info-icon"
          role="img"
        >
          ⚽
        </span>
      </div>
      <div
        class="column is-one-fifth match-info has-text-roma-red has-text-centered"
      >
        <p>
          <span
            class="match-info-score"
          >
            1
          </span>
          <span
            class="has-text-team-granite"
          >
            -
          </span>
          <span
            class="match-info-score"
          >
            2
          </span>
        </p>
      </div>
      <div
        class="column is-two-fifths team-info"
      >
        <span
          aria-label="Soccer ball"
          class="team-info-icon"
          role="img"
        >
          ⚽
        </span>
        <span
          class="team-info-abrv"
        >
          CMS2
        </span>
      </div>
    </section>
  </article>
</DocumentFragment>
`;

exports[`Event Card has not changed--match-future 1`] = `
<DocumentFragment>
  <article
    class="event-row-container container"
  >
    <header
      class="event-row-header"
    >
      <p>
        20 NOV 2200
      </p>
      <p
        class="has-text-weight-light"
      >
        Parco Sannazzaro
      </p>
    </header>
    <section
      class="event-row-info columns is-mobile is-vcentered"
    >
      <div
        class="column is-two-fifths has-text-right team-info"
      >
        <span
          class="team-info-abrv"
        >
          CMS
        </span>
        <span
          aria-label="Soccer ball"
          class="team-info-icon"
          role="img"
        >
          ⚽
        </span>
      </div>
      <div
        class="column is-one-fifth match-info has-text-roma-red has-text-centered"
      >
        <p>
          12:40 PM
        </p>
      </div>
      <div
        class="column is-two-fifths team-info"
      >
        <span
          aria-label="Soccer ball"
          class="team-info-icon"
          role="img"
        >
          ⚽
        </span>
        <span
          class="team-info-abrv"
        >
          CMS2
        </span>
      </div>
    </section>
  </article>
</DocumentFragment>
`;

exports[`Event Card has not changed--null 1`] = `
<DocumentFragment>
  <article
    class="event-row-container container"
    data-testid="event-row-empty"
  />
</DocumentFragment>
`;

exports[`Event Card has not changed--practice 1`] = `
<DocumentFragment>
  <article
    class="event-row-container container"
  >
    <header
      class="event-row-header"
    >
      <p>
        20 NOV 2019
      </p>
      <p
        class="has-text-weight-light"
      >
        Parco Sannazzaro
      </p>
    </header>
    <section
      class="event-row-info has-text-centered columns is-mobile is-vcentered"
    >
      <div
        class="column practice-info"
      >
        <p>
          PRACTICE
        </p>
        <p
          class="has-text-roma-red"
        >
          12:40 PM
        </p>
      </div>
    </section>
  </article>
</DocumentFragment>
`;
